IMX477 Improvements - 6.12

6by9 opened the pull request:
6.12 version of #6208 

~Only compile tested.~

ameybarapatre commented:
Hi, can we rebuild?, getting 404 error

6by9 commented:
Rebased which will retrigger the CI builds.

My comment from https://github.com/raspberrypi/linux/pull/6208#issuecomment-2161173009 that  2028x1080 and 2028x1520 modes with 10 bit readout give totally white frames needs to be resolved before contemplating merging.
Unless you're saying that you've tested those modes and that they work.

6by9 committed:
media: imx477: Remove duplicated writes to line_length_pix

Register line_length_pix was being written by both the tables
of registers and the control handler for V4L2_CID_HBLANK.

Remove the duplication in the tables.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: imx477: Convert line_length_pix values to decimal

line_length_pix is a value that the developer wants to know,
so write the values in decimal.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: imx477: Precompute frame length values

The frame length default value doesn't change dynamically, and
neither does any of the other parameters that configure it,
so precompute it instead of working from a frame duration to
get to the value.

The minimum value was also computed, when actually the sensor
will take any value down to 4 lines.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: imx477: Convert to use V4L2_CCI library

This removes a load of boilerplate code around how registers
are grouped into multiple word values.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: i2c: imx477: Extract more registers from mode tables to common

There are a fair number of registers duplicated in all the mode
tables, so move those into the common table.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: i2c: imx477: Add full res but cropped 16:9 mode

For 4k30 recording we want 16:9 output, so add a cropped mode
to achieve this.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: imx477: Compute link frequency PLL settings

Rather than the hard coded PLL settings for fixed frequencies,
compute the PLL settings based on device tree, validating that
the specified rate can be achieved.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: imx477: Compute line_length_pix based on link frequency

As we now support variable link frequency, compute the minimum
line_length value that the sensor will work with, and set
V4L2_CID_HBLANK based on that number.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
Thanks for testing.

As this is a sensor driver, I'm not that interested in anything above 750MHz as then you're exceeding the Pi's capabilities rather than the sensor's. 
Frame drops and corrupt frames are again most likely to be the Pi rather than the sensor. Ideally I'd like to get CSI2 level debug for error rates etc, but I don't believe I have an easy way to do that.

Pi5's libcamera pipeline handler already knows that RP1 CFE has a limit on pixel rate (250MPix/s IIRC), so I'm expecting that will be automatically extending the HBLANK control value once you get to a point where the link frequency can exceed the limit, but you won't get any increase in output framerate from libcamera on Pi5.

Carnarts commented:
@schoolpost Are you managing to record ok the uncropped 4056x3040/24 modes? Curious how you're managing that if so. Testing with Cinemate, the preview looks good but as soon as I record the preview drops frames making these modes unusable for me, it isn't dropping frames on the disk but at the camera.

schoolpost commented:
> @schoolpost Are you managing to record ok the uncropped 4056x3040/24 modes? Curious how you're managing that if so. Testing with Cinemate, the preview looks good but as soon as I record the preview drops frames making these modes unusable for me, it isn't dropping frames on the disk but at the camera.

I don't believe that would be driver related issue, more so application level and buffer management ( recycling buffers in a timely manner )

6by9 commented:
You're right - it's 380MPix/s that libcamera limits the pixel rate to. My memory is going!
https://github.com/raspberrypi/libcamera/blob/main/src/ipa/rpi/controller/controller.cpp#L62-L78

Greater than 750MHz is still above RP1's spec, so it's still debatable as to whether it is the sensor setup or Pi misbehaving.
IMX477 spec is up to 2.1Gbit/s/lane, so it should be able to send data up to 1050MHz. 

I haven't confirmed that the PLL configuration is still within the specs. Spec says the max multiplier value is 350.
....
A quick test says 924MHz uses a multiplier of 0x134 or 308. Scaling that linearly would give 1050MHz with a multiplier of 350.

The pixel rate PLL is always set to 840MHz, and that is sufficient for full res @60fps, so that shouldn't be a limit. Just as long as we set LINE_LENGTH_PCK to a sufficient value to be able to transmit all the data.

6by9 committed:
media: imx477: Write the REQ_LINK_BIT_RATE reg programmatically

Now that the link frequency can be varied, write the link bit
rate registers to reflect the speed being used.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: imx477: Use auto global timings if not at 450MHz link freq

The timing registers configured are for 450MHz.
If running at a different link frequency, use the automatic
timing control.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: imx477: Support 10 or 12 bit readout for all modes

The sensor supports readout as 10 or 12 bit. As we are now
computing the horizontal blanking limits dynamically, adding
support for both readout modes falls out trivially, so add
them both.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 committed:
media: imx477: Add support for 8 bit readout

8 bit readout is only a reconfiguration of the CSI2 block,
and recomputation of horizontal blanking. Enable it.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.com>

6by9 commented:
I'm done then. Over to @naushir to review.

fliker09 commented:
A stupid question - once this get merged, all these modes will be available in the standard 6.12 kernel, right? Will it still require dtoverlay to enable?

6by9 commented:
> A stupid question - once this get merged, all these modes will be available in the standard 6.12 kernel, right? Will it still require dtoverlay to enable?

8, 10, 12 bit readout for all modes will be available as standard even when the firmware autodetection is used.

The default link frequency will remain at 450MHz, giving
```
0 : imx477 [4056x3040 12-bit RGGB] (/base/axi/pcie@1000120000/rp1/i2c@80000/imx477@1a)
    Modes: 'SRGGB10_CSI2P' : 1332x990 [125.83 fps - (696, 528)/2664x1980 crop]
                             2028x1080 [77.77 fps - (0, 440)/4056x2160 crop]
                             2028x1520 [55.32 fps - (0, 0)/4056x3040 crop]
                             4056x2160 [19.98 fps - (0, 440)/4056x3040 crop]
                             4056x3040 [14.20 fps - (0, 0)/4056x3040 crop]
           'SRGGB12_CSI2P' : 1332x990 [106.18 fps - (696, 528)/2664x1980 crop]
                             2028x1080 [65.36 fps - (0, 440)/4056x2160 crop]
                             2028x1520 [46.49 fps - (0, 0)/4056x3040 crop]
                             4056x2160 [16.72 fps - (0, 440)/4056x3040 crop]
                             4056x3040 [11.89 fps - (0, 0)/4056x3040 crop]
           'SRGGB8' : 1332x990 [154.44 fps - (696, 528)/2664x1980 crop]
                      2028x1080 [96.02 fps - (0, 440)/4056x2160 crop]
                      2028x1520 [68.29 fps - (0, 0)/4056x3040 crop]
                      4056x2160 [24.82 fps - (0, 440)/4056x3040 crop]
                      4056x3040 [17.64 fps - (0, 0)/4056x3040 crop]
```
which is a small bump in most modes.

At least initially increasing the link frequency will require a manual `dtoverlay=imx477,link_frequency=N` in config.txt. You've seen from the above that most of my testing has been on Pi5 where the SoC specs officially go up to 750MHz. Increasing above 500MHz on Pi0-4 or 750MHz on Pi5 will be considered outside of spec, so support will be limited.

I'll discuss with colleagues as to whether it is worth modding the Pi5 bootloader to automatically add an override to 750MHz for autodetected cameras, thereby increasing the default capabilities.
(It may get pushed to a `pi5` override instead and do it to all autodetected cameras. #7057 is in progress adding a similar change to imx708, so we could have a performance increase on Pi5 there too).

6by9 commented:
> One more thing, just to get a reality check. Do I understand it correctly that encoding at 2160p in H.264 in real-time is unachievable even on Pi 5? From what I read on pi forums - CPU just won't be able to keep up... Is MJPEG the only viable option? (and I guess RAW dumping should work too)

2160p frames can be encoded as H264, but the frame rate achievable will depend on encoding options. Definitely no B-frames, potentially 1 reference frame, bitrate may be limited due to the complexities of CABAC.
MJPEG is far simpler from an encoding complexity, but you should get better encoding efficiency with H264 on all I-frames mode with minimal extra computational complexity.
